[Nvidia GPU] Add mechanism to detect nccl timeout and return error status

Tixxx opened the pull request:
The current behavior crashes the program whenever a nccl async error has occured, timeout errors are also not detected for async events. This pr adds a mechanism to do:
1. poll statuses of async events and return timeout if status is pending for too long
2. return nccl async event status as xla status so a proper python exception can be thrown.

Tixxx committed:
Add mechanism to detect nccl timeout and return error status

golechwierowicz commented:
@frgossen can you take a look?

frgossen commented on xla/service/gpu/runtime/thunk.h:
```diff
@@ -270,6 +270,10 @@ class Thunk {
     int64_t collective_max_nchannels;
     int64_t p2p_max_nchannels;
 
+    absl::Status async_status = absl::OkStatus();
```
nit: can you add a comment explaining the purpose of these two?

Tixxx replied:
Yes sure. Also I'm moving these to gpu_executable since it makes more sense for these to be part of the global executable. I should have another commit by EOD. @frgossen can we hold off on merging it today? i will temporarily change it to draft for now.

Tixxx replied:
moved the async queue and async status to gpu executable. please take another look. thanks.

frgossen reviewed (approved):
Thanks! One comment

Tixxx committed:
move async status and queue management to gpu executable

Tixxx commented:
@frgossen Could you take another look at this PR? Thanks

frgossen reviewed (approved):
Thanks!

frgossen reviewed (changes requested):
Do you think we can add a test case for this? We could trigger a NCCL deadlock in a test case

Tixxx commented:
> Do you think we can add a test case for this? We could trigger a NCCL deadlock in a test case

Artifacting a collective deadlock might not be easy with a simple test case. I have been trying to get an hlo with send/recv to deadlock but im hitting a bunch of errors in hlo verifier, looks like they have to go through some specific passes that add attributes for the compiler to not complain. I will try some more.

sgerrard commented:
@Tixxx Do you have an updated status?

Tixxx committed:
Added e2e test for testing nccl timeout and error propagation

Tixxx commented:
> @Tixxx Do you have an updated status?

I'm able to repro a deadlock using a recv program and add it to the e2e test suite, @frgossen could you take another look?

frgossen reviewed (changes requested):
Thanks for adding the test case. Added a few comments.

Tixxx commented on xla/tests/collective_ops_e2e_test.cc:
```diff
@@ -1103,5 +1103,68 @@ ENTRY entry {
+  }
+
+  auto result = HloTestBase::ExecuteReplicated(
+      std::move(module), fake_ptrs, kNumPartitions, &assn,
```
The interface for the ExecuteReplicated is a bit confusing, the num_replicas in the argument is how many threads(devices) it will run on, the actual partitioning is done using the model config, in this case it's 1 replica and 2 partitions. I have run into a bunch of check failures when using all replicas so had to resort to using multiple partitions.

Tixxx committed:
address pr comments

Tixxx commented on xla/tests/collective_ops_e2e_test.cc:
```diff
@@ -1103,5 +1103,68 @@ ENTRY entry {
+    recv-data.0 = u32[2] get-tuple-element(recv-done.0), index=0
+
+    recv-done.1 = (u32[2], token[]) recv-done(recv.1), channel_id=2
+    recv-data.1 = u32[2] get-tuple-element(recv-done.1), index=0
```
updated

Tixxx commented on xla/python/pjit.cc:
```diff
@@ -877,7 +877,11 @@ PyObject* PjitFunction_tp_vectorcall(PyObject* callable, PyObject* const* args,
     if (!out.ok()) {
-      PyErr_SetString(PyExc_ValueError, out.status().ToString().c_str());
+      if (out.status().code() == absl::StatusCode::kDeadlineExceeded) {
+        PyErr_SetString(PyExc_TimeoutError, out.status().ToString().c_str());
```
updated

Tixxx commented on xla/python/pjit.cc:
```diff
@@ -41,11 +41,11 @@ limitations under the License.
 #include "nanobind/stl/string_view.h"  // IWYU pragma: keep
-#include "nanobind/stl/vector.h"  // IWYU pragma: keep
+#include "nanobind/stl/vector.h"       // IWYU pragma: keep
 #include "xla/pjrt/exceptions.h"
```
yea, the formatter decided this was best way to format. i changed it back.

Tixxx replied:
well looks like i have to do the formatting here otherwise the clang-format step in the CI will fail.

Tixxx committed:
changed back the formatting for xla/python/pjit.cc

Tixxx commented:
@frgossen addressed your comments above, can you take another look?

frgossen reviewed (approved):
Thanks!

penpornk reviewed (changes requested):
Could you please help take a look at CI failures?

Also, when I ran the `CollectiveOpsTestE2E.NcclErrorPropagation` test internally, I got SIGSEGV (general-protection fault). Does your test require a specific driver version, etc?

Tixxx commented:
> Could you please help take a look at CI failures?
> 
> Also, when I ran the `CollectiveOpsTestE2E.NcclErrorPropagation` test internally, I got SIGSEGV (general-protection fault). Does your test require a specific driver version, etc?

Im using 535.161.07 cuda 12.5, also im running this on hopper. Which platform and cuda version did you observe the SIGSEGV?

Tixxx committed:
Fix ci failure

Tixxx committed:
Add a grace period when async error happens so nccl has time to shut
down

Tixxx commented:
> Could you please help take a look at CI failures?
> 
> Also, when I ran the `CollectiveOpsTestE2E.NcclErrorPropagation` test internally, I got SIGSEGV (general-protection fault). Does your test require a specific driver version, etc?

@penpornk I finally got a repro with a different nccl version. I think the issue is that when the gpu executable is shutting down, nccl hasn't properly torn down yet. I added a small delay before exiting the gpu executable. The test is passing for all platforms for me, can you give it another try?

Tixxx commented:
@frgossen @penpornk  i have addressed the segsev issue for the test. Can you guys give it a try and take another look at the pr? There hasn't been any progress for the past few days.

frgossen commented:
>  I added a small delay before exiting the gpu executable
I don't think we should rely on a small delay. This will almost certainly lead to flaky tests. Can you try to find out why exactly the issue appears?

akuegel commented on xla/service/gpu/gpu_executable.cc:
```diff
@@ -80,6 +80,7 @@ limitations under the License.
 #if GOOGLE_CUDA || TENSORFLOW_USE_ROCM
 #include "xla/stream_executor/gpu/gpu_activation.h"
 #include "xla/stream_executor/gpu/gpu_executor.h"
+#include "xla/stream_executor/gpu/gpu_stream.h"
```
This include needs a corresponding dependency in the BUILD target.

Tixxx replied:
updated

Tixxx commented:
>  I added a small delay before exiting the gpu executable
> > I don't think we should rely on a small delay. This will almost certainly lead to flaky tests. Can you try to find out why exactly the issue appears?

@frgossen  The root cause is that we are aborting nccl comms in a separate monitor thread, when the main thread shuts down the gpu executable, nccl comms are not all cleaned up yet. Since the test infra runs the same test multiple times for different platforms consecutively next to each other, subsequent runs will crash due to previous nccl comms are not cleaned up in time. We can do a refactoring to move the nccl comms aborting logic to gpu executable and wait for the nccl abort call to finish before exiting.

Tixxx committed:
Added missing include in bazel rule

Tixxx commented:
> It seems there is now a merge conflict. Can you please resolve this?

yes, just rebased

frgossen commented:
> @frgossen The root cause is that we are aborting nccl comms in a separate monitor thread, when the main thread shuts down the gpu executable, nccl comms are not all cleaned up yet. Since the test infra runs the same test multiple times for different platforms consecutively next to each other, subsequent runs will crash due to previous nccl comms are not cleaned up in time. We can do a refactoring to move the nccl comms aborting logic to gpu executable and wait for the nccl abort call to finish before exiting.

Sounds much cleaner indeed. Let's do that instead of waiting

Tixxx committed:
remove the delay before exiting gpu executable
wait for the signal instead

Tixxx commented:
> > @frgossen The root cause is that we are aborting nccl comms in a separate monitor thread, when the main thread shuts down the gpu executable, nccl comms are not all cleaned up yet. Since the test infra runs the same test multiple times for different platforms consecutively next to each other, subsequent runs will crash due to previous nccl comms are not cleaned up in time. We can do a refactoring to move the nccl comms aborting logic to gpu executable and wait for the nccl abort call to finish before exiting.
> 
> Sounds much cleaner indeed. Let's do that instead of waiting

@frgossen  i have done a slightly different change than doing refactoring, I'm using a boolean to indicate the status of commAbort of all communicators, the flag is used in gpu executable to wait for all comms to be aborted before exiting so we dont have to use a fixed delay to wait. This is a bit simpler than refactoring the whole checkComm logic.

Tixxx committed:
add back IsStreamIdle api

Tixxx commented:
@frgossen  can we expedite the review on this one? The APIs in stream executor keep changing and i have to change a number of places in cuda driver and executor to make this PR work as intended.

frgossen commented:
> @frgossen can we expedite the review on this one? The APIs in stream executor keep changing and i have to change a number of places in cuda driver and executor to make this PR work as intended.

I am having a hard time finding the relevant changes in this PR. I guess it is really a Gitgub issue but can you somehow remove all those unrelated formatting changes? 

> slightly different change than doing refactoring, I'm using a boolean to indicate the status 

Wouldn't the right way to do this be some kind of a lock?

Tixxx committed:
removed formatting changes in pjit.cc

Tixxx commented:
> > @frgossen can we expedite the review on this one? The APIs in stream executor keep changing and i have to change a number of places in cuda driver and executor to make this PR work as intended.
> 
> I am having a hard time finding the relevant changes in this PR. I guess it is really a Gitgub issue but can you somehow remove all those unrelated formatting changes?
Sure i have removed the formatting changes however the clang format would fail on github, i guess they need to be fixed during import.

> 
> > slightly different change than doing refactoring, I'm using a boolean to indicate the status
> 
> Wouldn't the right way to do this be some kind of a lock?
The flag is set to true after all communicators are aborted and NcclCommAbort is a synchronous call, using a light-weight flag should be sufficient here or do you see other circumstances where this would fail?

thomasjoerg commented:
@frgossen Please have another look.

derdrdirk commented:
Hi @Tixxx,
I saw that you a `#if GOOGLE_CUDA || TENSORFLOW_USE_ROCM` Macro. The XLA team is actively working to reduce the number of these macros in the codebase as they hurt maintainability.

Would it be possible to remove the Macro?

---

```
# gpu_executable.cc
#if GOOGLE_CUDA || TENSORFLOW_USE_ROCM
    while (!se::gpu::AsGpuStream(stream_to_sync)->IsIdle()) {
      if (!async_status.async_op_status.ok()) {
        // NCCL error has occurred, wait for all communicators
        // to be aborted before returning.
        while (async_status.is_all_comms_aborted == false) {
          // Need to rendezvous while waiting for the signal in case
          // any rank is stuck when aborting.
          TF_RETURN_IF_ERROR(
              RendezvousAfterInitialization(run_options, debug_options));
        }

        return async_status.async_op_status;
      }
    }
#endif  // GOOGLE_CUDA || TENSORFLOW_USE_ROCM
```

derdrdirk commented:
If you want to get rid of the Macro maybe you can use something like in `gpu_executor.cc`

```
auto device = executor->GetDeviceDescription();
    // #if GOOGLE_CUDA || TENSORFLOW_USE_ROCM
    if (std::holds_alternative<se::CudaComputeCapability>(
            device.gpu_compute_capability()) ||
        std::holds_alternative<se::RocmComputeCapability>(
            device.gpu_compute_capability())) {
```

I tried to do it myself, but the PR got out of sync with head and I don't know the right way to "readd" the `IsIdle` method you added to the stream.

Tixxx committed:
Added stream idle check in cuda stream class

Tixxx commented:
> If you want to get rid of the Macro maybe you can use something like in `gpu_executor.cc`
> 
> ```
> auto device = executor->GetDeviceDescription();
>     // #if GOOGLE_CUDA || TENSORFLOW_USE_ROCM
>     if (std::holds_alternative<se::CudaComputeCapability>(
>             device.gpu_compute_capability()) ||
>         std::holds_alternative<se::RocmComputeCapability>(
>             device.gpu_compute_capability())) {
> ```
> 
> I tried to do it myself, but the PR got out of sync with head and I don't know the right way to "readd" the `IsIdle` method you added to the stream.

Thanks @derdrdirk , there's been frequent changes in those stream classes and it's been troublesome for me to catch the daily changes.
I removed the macro in gpu executor and added the IsIdle api to the stream interface and cuda stream class.

olegshyshkov commented:
@frgossen Are there any changes required on this PR before we can merge?

sanjoy reviewed (commented):
Minor nit inline.

Tixxx committed:
use a temp to hold the status string

Tixxx commented on xla/python/pjit.cc:
```diff
@@ -952,7 +952,12 @@ PyObject* PjitFunction_tp_vectorcall(PyObject* callable, PyObject* const* args,
         o->fun.Call(callable, args, nargs, kwnames);
     if (!out.ok()) {
-      PyErr_SetString(PyExc_ValueError, out.status().ToString().c_str());
+      const char* error_string = out.status().ToString().c_str();
```
ah thanks for pointing it out, i overlooked this. The jax models are able to run with this so i think clang might have detected it and created a temp. I added a temp to hold the string before getting the c string out of it.

jprabhas replied:
Would this sort of invalid mem ptr issues be caught by Google's internal CI. Does it run address sanitizers on the modified code? @frgossen ?

Tixxx commented:
> More than 10 different tests from //third_party/py/jax/experimental/jax2tf/tests/... failed.
> 
> For example //third_party/py/jax/experimental/jax2tf/tests:call_tf_graph_serialization_test_gpu_pjrt_c_api

Thanks. Let me try to repro this from my side.

Tixxx commented:
i fixed the failing tests but the recent collectives API refactoring caused a lot of conflicts. Im working to refit the pr to the new collectives API, converting this to draft for now

frgossen commented:
Is this PR still active or should we close it?
